sql-streaming-copier/README.md

```diff
@@ -0,0 +1,127 @@
+# Pub/Sub Copy Pipeline
+
+A dataflow template which performs copies to and from streaming systems using
+beam SQL.
+
+This can perform copies from any of:
+1) Cloud Pub/Sub
+2) Pub/Sub Lite
+3) Kafka
+
+This can perform copies to any of:
+1) Cloud Pub/Sub
+2) Pub/Sub Lite
+3) Kafka
+4) BigQuery
+
+All sources and sinks should have a simple SQL transform to/from the following
+representation, supported natively by Pub/Sub Lite and Kafka:
+
+```sql
+CREATE EXTERNAL TABLE <tabletype>(
+    message_key BYTES,
+    event_timestamp TIMESTAMP,
+    attributes ARRAY<ROW<key VARCHAR, values ARRAY<BYTES>>>,
+    payload BYTES,
+)
+```
+
+If copying to Pub/Sub, all attribute values must be interpretable as UTF-8
+strings. If multiple values are provided for an attribute, they will be joined
+with commas in the Pub/Sub Message.
+
+If copying to BigQuery, the table must accept the above format if it already
+exists.
+
+It requires the following options:
+
+* sourceType: [`pubsub`, `pubsublite`, `kafka`]
+* sourceLocation:
+  * `pubsub`: The topic (discouraged) or subscription to read from
+  * `pubsublite`: The subscription to read from
+  * `kafka`: The broker URL to read from
+* sinkType: [`pubsub`, `pubsublite`, `kafka`, `bigquery`]
+* sinkLocation:
+  * `pubsub`: The topic to write to
+  * `pubsublite`: The topic to write to
+  * `kafka`: The broker URL to write to
+  * `bigquery`: Location of the table in the BigQuery CLI format
+  
+When running with kafka, you must also set the following parameters:
+
+```bash
+--parameters='[source/sink]Options=^NOT_IN_JSON^{"topics":["<kafka topic name>"],"bootstrap.servers":"<broker URL>"}'
```

Comment:
I'm not sure I follow this part; what does NOT_IN_JSON mean? Is that part of the actual CLI argument? Also, where is one supposed to include this in the gcloud command below?

Reply:
It is, its to do escaping. The gcloud CLI tries to be too magic and parse this as a list (since it includes a comma) as opposed to just passing it on. `NOT_IN_JSON` is the replacement for the comma token it usually uses for list separation.